edit/edcore/insert.go

```diff
@@ -133,136 +136,160 @@ func (ed *editor) commandStart() {
+func (ed *editor) applyMove(move func(string, int) int) func() {
+	return func() {
+		ed.dot = move(ed.buffer, ed.dot)
+	}
 }
 
-// NOTE(xiaq): A word is a run of non-space runes. When killing a word,
-// trimming spaces are removed as well. Examples:
-// "abc  xyz" -> "abc  ", "abc xyz " -> "abc  ".
+func (ed *editor) applyKill(move func(string, int) int) func() {
+	return func() {
+		index := move(ed.buffer, ed.dot)
 
-func (ed *editor) killWordLeft() {
-	if ed.dot == 0 {
-		return
-	}
-	space := strings.LastIndexFunc(
-		strings.TrimRightFunc(ed.buffer[:ed.dot], unicode.IsSpace),
-		unicode.IsSpace) + 1
-	ed.buffer = ed.buffer[:space] + ed.buffer[ed.dot:]
-	ed.dot = space
-}
-
-// NOTE(xiaq): A small word is either a run of alphanumeric (Unicode category L
-// or N) runes or a run of non-alphanumeric runes. This is consistent with vi's
-// definition of word, except that "_" is not considered alphanumeric. When
-// killing a small word, trimming spaces are removed as well. Examples:
-// "abc/~" -> "abc", "~/abc" -> "~/", "abc* " -> "abc"
-
-func (ed *editor) killSmallWordLeft() {
-	left := strings.TrimRightFunc(ed.buffer[:ed.dot], unicode.IsSpace)
-	// The case of left == "" is handled as well.
-	r, _ := utf8.DecodeLastRuneInString(left)
-	if isAlnum(r) {
-		left = strings.TrimRightFunc(left, isAlnum)
-	} else {
-		left = strings.TrimRightFunc(
-			left, func(r rune) bool { return !isAlnum(r) })
+		if index < ed.dot { // delete left
+			ed.buffer = ed.buffer[:index] + ed.buffer[ed.dot:]
+			ed.dot = index
+		} else { // delete right
+			ed.buffer = ed.buffer[:ed.dot] + ed.buffer[index:]
+		}
 	}
-	ed.buffer = left + ed.buffer[ed.dot:]
-	ed.dot = len(left)
 }
 
 func isAlnum(r rune) bool {
 	return unicode.IsLetter(r) || unicode.IsNumber(r)
 }
 
-func (ed *editor) killRuneLeft() {
-	if ed.dot > 0 {
-		_, w := utf8.DecodeLastRuneInString(ed.buffer[:ed.dot])
-		ed.buffer = ed.buffer[:ed.dot-w] + ed.buffer[ed.dot:]
-		ed.dot -= w
-	} else {
-		ed.flash()
-	}
-}
+func moveDotLeft(buffer string, dot int) int {
+	_, w := utf8.DecodeLastRuneInString(buffer[:dot])
+	return dot - w
+}
+
+func moveDotRight(buffer string, dot int) int {
+	_, w := utf8.DecodeRuneInString(buffer[dot:])
+	return dot + w
+}
+
+// Returns a move- function that moves the cursor to the beginning of
+// the nearest "run" (characters of the same category) to the left.
+//
+// The categorization of characters is derived from the categorizer
+// function, which returns for each character an integer representing
+// the category of the integer.  The return value `0` of the
+// categorizer function is used as a special category representing
+// non-word, or "always-skip" characters.  All other return values are
+// equivalent, used only to distinguish between different character
+// classes.
+//
+// See `categorizeWord`, `categorizeSmallWord`, and
+// `categorizeAlphanumeric` for examples.
+func moveDotLeftCategoryFunc(categorize func(rune) int) func(string, int) int {
```

Comment:
Can we make this a low-order function `moveDotLeftCategoryFunc(categorize func(rune) int, buffer string, dot int) int`? This will make the definition of the `moveDotLeft*` functions slightly more complex (did you know that `gofmt` actually doesn't mind if you define a series functions as one-liners without empty lines between them?), but it is slightly clearer IMO. Ditto for the `moveDotRightCategoryFunc`.